app/models/internal/scheduling/Job.java

```diff
@@ -41,9 +42,9 @@
     /**
      * Gets an <a href="https://en.wikipedia.org/wiki/HTTP_ETag">ETag</a> that changes each time the job changes.
      *
-     * @return The ETag.
+     * @return The ETag, if any.
      */
-    String getETag();
+    Optional<String> getETag();
```

Comment:
Right now ETags are computed in [models.ebean.Report::toInternal](https://github.com/ArpNetworking/metrics-portal/blob/master/app/models/ebean/Report.java#L203) since they use data model specific fields such as  `lastModified`. This creates a problem where a view model cannot be mapped into an internal model and persisted without an etag - but an etag cannot be computed without already having persistence metadata.

To resolve this dependency I just went ahead and codified the behavior - we don't have an etag in the view model case so the model shouldn't mark it as required. This means cached jobs will always reload if there is no etag to compare. Another possibility was bundling the etag with the Repository query methods - since those are really the only locations where we should have a tag - but those changes would have been much more widespread.

Reply:
Just so we're clear it's not that the existing models have etag; because they don't. It seems more a matter that the etag pattern was never completely flushed out. In fact, etags appear to only exist to represent the state of a collection of objects for the organization (which I don't think are themselves fully utilized).

We should be using the version for concurrent modification control on database backed entities. However, Report does not have a version column. The version in the database case can also be used as the basis of the etag (e.g. together with id and org id; ideally, hashed in some way to obfuscate the data). However, it's debatable whether we should do that (use repository specific fields in the etag computation).

If we compute the etag from private database fields then it must be on the internal model and must be optional. The problem with this is that it can lead to brittle logic where two absent etag values are considered equal (this code review does not make the mistake; but it could be a source of problems). Alternatively, we relegate the etag to the view layer and compute it from the contents (perhaps even from the view model). This goes a bit against the current design where etags are surface from the repositories (although that may not be correct).

Given that we already compute the etags at the repository level, I think it's safe to continue that until we need to refactor the pattern. Under that plan, I think you're solution is the correct one. The only change I would like to see is to add version and use that to compute the etag. However, this change can be done in a separate code review as this review is already rather lengthy the it would address functionality already committed to master.

Reply:
It's confusing to me to state that they don't have etags when they're present as fields of the class but I understand what you're saying. I think of etags as additional functionality that wraps a type, so I think we're in agreement when you say _"etags appear to only exist to represent the state of a collection of objects for the organization"_; they're an optimization.

> Under that plan, I think you're solution is the correct one. The only change I would like to see is to add version and use that to compute the etag.

Sure, just file an Issue/task so we can track that.